src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        stateCopy = removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+    }
+
+    states.push({ ...stateCopy });
+  }
+
+  return states;
+}
+
+function removeProperties(state, keysToRemove) {
+  const updatedState = { ...state };
+
+  for (const key of keysToRemove) {
+    delete updatedState[key];
+  }
+
+  return updatedState;
```

Comment:
The `removeProperties` function correctly removes specified keys from the state. However, ensure that `keysToRemove` is always an array to avoid runtime errors. Consider adding a check or validation for `keysToRemove` to ensure it is an array before iterating over it.